sewer/auth.py

```diff
@@ -1,42 +1,115 @@
+    each item in the authprizations list.  So providers that have to deal with one
+    auth at a time can just implement that.  Some providers may be able to work
+    more efficently if they can see all the items at once, and those will prefer
+    to implement *_cert, overriding the versions here.
+
+    __init__ arguments protocol:  Taking the lesson from some Django classes that
+    were layered like this (and with many leaf users, some outside of the framework),
+    it is always declared with only self and **kwargs.  Each layer MUST pop the args
+    it knows and wishes to handle, then pass what's left to its immediate base class.
+    The end goal is to allow initializing a provider with whatever it needs to
+    authenticate to and interact with the DNS system that publishes the challenges.
+    Then that can be done outside of sewer's ACME engine (Client class) and passed
+    to it ready to use.
+    """
+
+    def __init__(self, **kwargs):
         """
-        called after the cert is aquired to cleanup any authorization resources created in the fulfill_authorization method above.
+        chal_types      Sequence[str] set of implemented ACME challenge types
+
+        optional logging control argument(s).  From preferred to legacy:
+         *  logger	a configured logging object to use as-is
+         *  log_level	level to setup local logger for, defaulting to INFO
+         *  LOG_LEVEL   obsolescent legacy version of log_level
+
+        logger will override log_level will override LOG_LEVEL if more than one is passed
         """
-        raise NotImplementedError("cleanup_authorization method must be implemented.")
+
+        ### FIX ME ### auth_type handling?  and that's not best named, should be challenge_type
+
+        self.chal_types = kwargs.pop("chal_types")
+        if not all(ct in supported_validations for ct in self.chal_types):
+            raise ValueError("Unsupported challenge type in list: %s" % self.chal_types)
+
+        logger = kwargs.pop("logger", None)
+        log_level = kwargs.pop("LOG_LEVEL", "INFO")  # compatibility / obsolescent
+        log_level = kwargs.pop("log_level", log_level)
+        if kwargs:
+            raise ValueError("BaseAuth was passed unrecognized argument(s): %s" % kwargs)
+
+        if logger:
+            self.logger = logger
+        else:
+            self.logger = logging.getLogger(__name__)
+            self.logger.setLevel(log_level)
+            if not self.logger.hasHandlers():
+                handler = logging.StreamHandler()
+                formatter = logging.Formatter("%(message)s")
+                handler.setFormatter(formatter)
+                self.logger.addHandler(handler)
+
+    def setup_cert(self, authorizations: Sequence[Tuple[str, str, str]]):
+        for auth in authorizations:
+            self.setup_auth(*auth)
+
+    def is_ready_cert(self, authorizations: Sequence[Tuple[str, str, str]]):
```

Comment:
what is you plan for this method, and `is_ready_auth`? I'm not seeing them used in the client

Reply:
This was a late thought.  Use is obvious - after setup_cert, call is_ready_cert and driver MAY verify that the challenge data is in place... this is more an issue for some dns services, and when it's worth checking and how to check are best left to the driver.

Thanks for looking it over.  I'm going to finish this set of changes RealSoonNow... time just keeps slipping away from me. :-(

Reply:
ah, I see.

exited to see your changes, thanks for the work on this!